diff --git a/packages/yarnpkg-fslib/sources/MountFS.ts b/packages/yarnpkg-fslib/sources/MountFS.ts
index 655a0007..e11eb236 100644
--- a/packages/yarnpkg-fslib/sources/MountFS.ts
+++ b/packages/yarnpkg-fslib/sources/MountFS.ts
@@ -997,7 +997,7 @@ export class MountFS<MountedFS extends MountableFS> extends BasePortableFakeFS {
           continue;
 
         try {
-          if (this.typeCheck !== null && (this.baseFs.lstatSync(filePath).mode & constants.S_IFMT) !== this.typeCheck) {
+          if (this.typeCheck !== null && (this.baseFs.statSync(filePath).mode & constants.S_IFMT) !== this.typeCheck) {
             this.notMount.add(filePath);
             continue;
           }
